src/zcash/IncrementalMerkleTree.hpp

```diff
@@ -57,25 +58,36 @@ class MerklePath {
 template<size_t Depth, typename Hash>
 class EmptyMerkleRoots {
 public:
-    EmptyMerkleRoots() {
-        empty_roots.at(0) = Hash::uncommitted();
-        for (size_t d = 1; d <= Depth; d++) {
-            empty_roots.at(d) = Hash::combine(empty_roots.at(d-1), empty_roots.at(d-1), d-1);
-        }
+    void initialize() {
+        if (!initialized) {
+     LOCK(cs_EmptyMerkleRoots_private_variables_write);
+          if (!initialized) {
+            empty_roots.at(0) = Hash::uncommitted();
+            for (size_t d = 1; d <= Depth; d++) {
+                empty_roots.at(d) = Hash::combine(empty_roots.at(d-1), empty_roots.at(d-1), d-1);
+            }
+            initialized = true;
+          }
+       }
     }
+    EmptyMerkleRoots() { initialized = false; }
     Hash empty_root(size_t depth) {
+        initialize();
         return empty_roots.at(depth);
     }
     template <size_t D, typename H>
     friend bool operator==(const EmptyMerkleRoots<D, H>& a,
                            const EmptyMerkleRoots<D, H>& b);
 private:
+    CCriticalSection cs_EmptyMerkleRoots_private_variables_write;
+    bool initialized;
```

Comment:
I think this would be better as a `std::atomic<bool>`. We could then replace the locking with `initialized.compare_exchange_strong(...)`.

Reply:
```
template<size_t Depth, typename Hash>
class EmptyMerkleRoots {
public:
    EmptyMerkleRoots() : initialized(false) { }
    void initialize() const {
        bool unitialized = false;
        if (initialized.compare_exchange_strong(unitialized, true)) {
            empty_roots.at(0) = Hash::uncommitted();
            for (size_t d = 1; d <= Depth; d++) {
                empty_roots.at(d) = Hash::combine(empty_roots.at(d-1), empty_roots.at(d-1), d-1);
            }
        }
    }
    Hash empty_root(size_t depth) const {
        initialize();
        return empty_roots.at(depth);
    }
    template <size_t D, typename H>
    friend bool operator==(const EmptyMerkleRoots<D, H>& a,
                           const EmptyMerkleRoots<D, H>& b);
private:
    mutable std::atomic<bool> initialized;
    mutable std::array<Hash, Depth+1> empty_roots;
};
```

Reply:
This is a neat way to do it I guess. I'm not sure that it's so much better than what's proposed that we should hold up the PR though. Maybe someone else disagrees.